stake-pool/single-pool/src/processor.rs

```diff
@@ -0,0 +1,1079 @@
+        let pre_pool_stake = get_active_stake_state(pool_stake_info, clock.epoch)?
+            .delegation
+            .stake;
+        let pre_pool_lamports = pool_stake_info.lamports();
+        msg!("Stake pre merge {}", pre_pool_stake);
+
+        let pre_user_stake = get_active_stake_state(user_stake_info, clock.epoch)?
+            .delegation
+            .stake;
+        let user_unstaked_lamports = user_stake_info
+            .lamports()
+            .checked_sub(pre_user_stake)
+            .ok_or(SinglePoolError::ArithmeticUnderflow)?;
+
+        // merge the user stake account, which is preauthed to us, into the pool stake account
+        // this merge succeeding implicitly validates authority/lockup of the user stake account
+        Self::stake_merge(
+            vote_account_address,
+            user_stake_info.clone(),
+            pool_authority_info.clone(),
+            bump_seed,
+            pool_stake_info.clone(),
+            clock_info.clone(),
+            stake_history_info.clone(),
+            stake_program_info.clone(),
+        )?;
+
+        let post_pool_stake = get_active_stake_state(pool_stake_info, clock.epoch)?
+            .delegation
+            .stake;
+        let post_pool_lamports = pool_stake_info.lamports();
+        msg!("Stake post merge {}", post_pool_stake);
+
+        // all lamports added, as a lamport difference, both stake and excess rent
+        let lamports_added = post_pool_lamports
+            .checked_sub(pre_pool_lamports)
+            .ok_or(SinglePoolError::ArithmeticUnderflow)?;
+
+        // stake lamports added, as a stake difference
+        let stake_added = post_pool_stake
+            .checked_sub(pre_pool_stake)
+            .ok_or(SinglePoolError::ArithmeticUnderflow)?;
+
+        // excess rent, as all lamports less stake
+        let leftover_rent = lamports_added
+            .checked_sub(stake_added)
+            .ok_or(SinglePoolError::ArithmeticUnderflow)?;
+
+        // sanity check: our calculated new stake matches the user prior stake
+        if stake_added != pre_user_stake {
+            return Err(SinglePoolError::UnexpectedMathError.into());
+        }
+
+        // sanity check: our calculated excess rent matches the user prior rent
+        if leftover_rent != user_unstaked_lamports {
+            return Err(SinglePoolError::UnexpectedMathError.into());
+        }
+
+        // sanity check: the user stake account is empty
+        if user_stake_info.lamports() != 0 {
+            return Err(SinglePoolError::UnexpectedMathError.into());
+        }
+
+        // we add initial lamports to supply to make the math work without minting tokens to incinerator
+        let token_supply = {
+            let pool_mint_data = pool_mint_info.try_borrow_data()?;
+            let pool_mint = Mint::unpack_from_slice(&pool_mint_data)?;
+            pool_mint.supply.saturating_add(INITIAL_LAMPORTS)
```

Comment:
If the pool account may start with additional lamports, or if the minimum delegation amount is raised after pool creation, does this logic break? I think it's probably ok to use 1 SOL or the minimum delegation floor, but I haven't thought fully through the implications.

Reply:
we use the true supply now that i mint the initialization tokens as part of supporting minimum delegation